File: solutions/_740_delete_and_earn/solution.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var deleteAndEarn = function (nums) {
  if (nums.length === 0) {
    return 0;
  }

  const points = new Array(Math.max(...nums) + 1).fill(0);
  for (const num of nums) {
    points[num] += num;
  }

  let prev = 0;
  let curr = 0;
  for (let i = 0; i < points.length; i++) {
    const temp = curr;
    curr = Math.max(curr, prev + points[i]);
    prev = temp;
  }

  return curr;
};

module.exports = { deleteAndEarn };

/*
https://leetcode.com/problems/delete-and-earn/
*/

File: solutions/_740_delete_and_earn/solution.java

package _740_delete_and_earn;

import java.util.Arrays;

class Solution {
    public int deleteAndEarn(int[] nums) {
        int max = Arrays.stream(nums).max().getAsInt();
        int[] points = new int[max + 1];
        for (int num : nums) {
            points[num] += num;
        }

        int prev = 0;
        int curr = 0;
        for (int i = 0; i < points.length; i++) {
            int temp = curr;
            curr = Math.max(curr, prev + points[i]);
            prev = temp;
        }
        return curr;
    }
}

/*
 * https://leetcode.com/problems/delete-and-earn/
 */